### ISSUE TITLE
Doc Feature Request: For tf 2, explain how to distribute a pretrained model across multiple nodes.

### ISSUE THREAD

**Issue by arivero** [tensorflow/tensorflow#60270]
When I read the original TF paper, I was told that

> Dataflow simplifies distributed execution, because it makes communication between subcomputations explicit.
> It enables the same TensorFlow program to be deployed to a cluster of GPUs for training, a cluster of TPUs for serving, and a cellphone for mobile inference. 
> Each operation resides on a particular device, such as a CPU or GPU in a particular task. A device is responsible for executing a kernel for each operation assigned to it.
> (...)
> The TensorFlow runtime places operations on devices, subject to implicit or explicit constraints in the graph.
> The placement algorithm computes a feasible set of devices for each operation, calculates the sets of operations that must be colocated, and selects a satisfying device for each colocation group.

I work with small models, so I have never need more than one device. But now large models are all the rage, and a lot of competition is about just getting the weights of a pretrained model and run them. So I want cluster all the GPUs of a room to just run a LLM, either ad-hoc, keras or HF. 

I would expect a tutorial for this to be available somewhere. It is not. I guess one still needs to configure a cluster description and, if such option is available still in 2.x, launch an script in each machine to join them into a cluster. And then somehow launch a master script in some machine that loads the model, deploys its graph across the cluster, and runs it. Without needing any mirroring or any parameter collection, so that the current objects in distribute.strategy seem very big for such task. Perhaps eventually the example could include some ClusterResolver for common cloud services, but ideally it should be for in-home heterogeneous architecture.

Of course it could be still sensible to do such tutorial with the distribute.strategy tools, if it were considered as a first introduction to distribution of tasks, and that the next natural thing is to learn also to do a loss calculation, a backpropagation, and a training step.

**Comment by Adesoji1** [comment]
@arivero 

Running a large language model on a cluster of GPUs can be achieved with TensorFlow's distribution strategies. Since TensorFlow 2.x provides a much easier interface to work with, I'll outline a high-level approach for you to follow using TensorFlow 2.x, Keras, and Horovod.

Install the necessary packages:
Install TensorFlow 2.x.
Install Horovod, a distributed deep learning framework, which works well with TensorFlow, Keras, and other DL frameworks.
Set up the cluster:
For an in-house cluster of GPUs, make sure all devices are connected to the same network and have the necessary GPU drivers and software installed (CUDA, cuDNN, etc.).
For a cloud-based cluster, use the provider's GPU-based instances and set up a virtual private network (VPN) between the instances.
Configure TensorFlow and Horovod:
Import TensorFlow, Keras, and Horovod.
Initialize Horovod by calling horovod.tensorflow.keras.init().
Prepare your model:
Load your large language model using TensorFlow or Keras, e.g., tf.keras.models.load_model('path/to/model').
Wrap the model's optimizer with horovod.DistributedOptimizer(optimizer).
Compile the model with the wrapped optimizer and appropriate loss and metric functions.
Configure the training and evaluation steps:
Set up your dataset and data pipeline.
Wrap your dataset with tf.data.experimental.parallel_interleave() or tf.data.experimental.CsvDataset.from_generator() to distribute the data across multiple GPUs.
Define a custom training loop or use the built-in model.fit() and model.evaluate() functions. With Horovod, the training and evaluation steps will be automatically distributed across the available GPUs.
Run the training and evaluation:
Use horovodrun to launch your script on all the available devices. For example, horovodrun -np 4 -H localhost:4 python your_script.py will run your script on 4 GPUs.
This high-level approach should provide a starting point for running a large language model on a cluster of GPUs using TensorFlow 2.x, Keras, and Horovod. You may need to fine-tune the configuration and code to adapt to your specific setup and requirements.

**Comment by arivero** [comment]
@Adesoji1 just to be sure can you confirm that it I use a model such that **it does not fit in one GPU** but its graph can be put across four GPUs, then horovod allows me to run inference? I can not see in the documentation any decent sample of model parallelism, not even GPUs in the same machine, and we are speaking of multi-node clusters.

Even if it does, I still think that a tensorflow only tutorial would be a good thing, as distributing graphs was one of the design goals of tensorflow, so it could be good to exhibit this capability.

**Comment by SuryanarayanaY** [comment]
Hi @arivero ,

For training on multiple works you can go through the attached tutorials [here](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras) and [here](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_ctl).All these uses data parallelism concept.

However if you are looking for model parallelism right now Tensorflow not supporting model parallelism.Please refer to the source [here](https://www.tensorflow.org/api_docs/python/tf/distribute#:~:text=Data%20parallelism%20is,in%20the%20future).Here Model parallelism refers to splitting the Model on to different Ops and placing them on to different devices/workers and executing the Ops in Parallel. Right now Tensorflow supports data Parallelism i.e. splitting the data into different devices/workers.

If I am still missing something here please let me know. Thanks!

**Comment by arivero** [comment]
Thanks, yes, I was looking for model parallelism, I see that it is not supported for training, but I was hoping that some support for inference only, when the graph is already know and trained. There is some legacy information around where it is suggested to use submodels and tf.device() contexts, to deploy manually, but it is mainly for tensorflow 1.x and not sure if it really works across servers.

What I was looking was a recipe such as:
- build the model.
- from the model, obtain the graph.
- for the graph, loop manually across, using tf.device() or something similar to allocate each operation to a device.
- execute.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
